rest/blip_api_test.go

```diff
@@ -4494,3 +4494,82 @@ func TestBlipLegacyAttachDocUpdate(t *testing.T) {
 		require.Equal(t, keyValueErr.DocumentID, v2Key)
 	}
 }
+
+// Tests changes made in CBG-2151 to return errors from sendRevision unless it's a document not found error,
+// in which case a noRev should be sent.
+func TestSendRevisionNoRevHandling(t *testing.T) {
+	if !base.UnitTestUrlIsWalrus() {
+		t.Skip("Skip LeakyBucket test when running in integration")
+	}
+	testCases := []struct {
+		name        string
+		error       error
+		expectNoRev bool
+	}{
+		{
+			name:        "not_found",
+			error:       gocb.ErrDocumentNotFound,
+			expectNoRev: true,
+		},
+		{
+			name:        "overload",
+			error:       gocb.ErrOverload,
+			expectNoRev: false,
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			rt := NewRestTester(t, &RestTesterConfig{
+				guestEnabled: true,
+				TestBucket:   base.GetTestBucket(t).LeakyBucketClone(base.LeakyBucketConfig{}),
+			})
+			defer rt.Close()
+
+			leakyBucket, ok := base.AsLeakyBucket(rt.Bucket())
+			require.True(t, ok)
+
+			btc, err := NewBlipTesterClientOptsWithRT(t, rt, nil)
+			require.NoError(t, err)
+			defer btc.Close()
+
+			// Change noRev handler so it's known when a noRev is received
+			recievedNoRevs := make(chan *blip.Message)
+			btc.pullReplication.bt.blipContext.HandlerForProfile[db.MessageNoRev] = func(msg *blip.Message) {
+				fmt.Println("Received noRev")
+				recievedNoRevs <- msg
+			}
+
+			resp := rt.SendAdminRequest(http.MethodPut, "/db/"+test.name, `{"foo":"bar"}`)
+			assertStatus(t, resp, http.StatusCreated)
+
+			// Make the LeakyBucket return an error
+			leakyBucket.SetGetRawCallback(func(key string) error {
+				return test.error
+			})
+
+			// Flush cache so document has to be retrieved from the leaky bucket
+			rt.GetDatabase().FlushRevisionCacheForTest()
+
+			err = btc.StartPull()
+			assert.NoError(t, err)
+
+			// Wait 3 seconds for noRev to be received
+			select {
+			case msg := <-recievedNoRevs:
+				if test.expectNoRev {
+					assert.Equal(t, test.name, msg.Properties["id"])
```

Comment:
I'd change this to require, because the subsequent assertion on 4572 might end up being true if NoRev happens.

Reply:
Left that as an assert because if it receives a noRev for a different document (which should never happen), then it would be interesting to know if the actual rev does get successfully replicated (which it shouldn't). Does this sound ok?